bin/Macronify/MacroniRewriters.cpp

```diff
@@ -0,0 +1,281 @@
+                    type = ty_attr;
+                }
+                member = member_op.getNameAttr();
+            }});
+        auto is_legal = ptr && type && member;
+        if (!is_legal) {
+            return mlir::failure();
+        }
+
+        auto op = exp.getOperation();
+        auto ptr_clone = rewriter.clone(*ptr);
+        auto exp_ty = exp.getType(0);
+        auto ptr_res = ptr_clone->getResult(0);
+
+        using CO = ::macroni::kernel::ContainerOf;
+        rewriter.replaceOpWithNewOp<CO>(op, exp_ty, ptr_res, *type, *member);
+        return mlir::success();
+    }
+
+    mlir::LogicalResult rewrite_rcu_dereference(
+        macroni::MacroExpansion exp,
+        mlir::PatternRewriter &rewriter) {
+        if (!is_rcu_dereference(exp)) {
+            return mlir::failure();
+        }
+        mlir::Operation *p = nullptr;
+        exp.getExpansion().walk([&](mlir::Operation *op) {
+            if (auto param = mlir::dyn_cast<macroni::MacroParameter>(op)) {
+                auto param_name = param.getParameterName();
+                if (param_name == "p") {
+                    p = op;
+                }
+            }});
+        auto is_legal = p != nullptr;
+        if (!is_legal) {
+            return mlir::failure();
+        }
+
+        auto op = exp.getOperation();
+        auto p_clone = rewriter.clone(*p);
+        auto res_ty = exp.getType(0);
+        auto p_res = p_clone->getResult(0);
+        using RCUD = ::macroni::kernel::RCUDereference;
+        rewriter.replaceOpWithNewOp<RCUD>(op, res_ty, p_res);
+        return mlir::success();
+    }
+
+    mlir::LogicalResult rewrite_smp_mb(
+        macroni::MacroExpansion exp,
+        mlir::PatternRewriter &rewriter) {
+        if (!is_smp_mb(exp)) {
+            return mlir::failure();
+        }
+
+        auto op = exp.getOperation();
+        auto res_ty = exp.getType(0);
+        rewriter.replaceOpWithNewOp<kernel::SMPMB>(op, res_ty);
+        return mlir::success();
+    }
+
+    mlir::LogicalResult rewrite_list_for_each(
+        vast::hl::ForOp for_op,
+        mlir::PatternRewriter &rewriter) {
+        mlir::Operation *pos = nullptr;
+        mlir::Operation *head = nullptr;
+        for_op.getCondRegion().walk(
+            [&](mlir::Operation *op) {
+                using MP = macroni::MacroParameter;
+                if (auto param_op = mlir::dyn_cast<MP>(op)) {
+                    if (param_op.getParameterName() == "pos") {
+                        pos = op;
+                    } else if (param_op.getParameterName() == "head") {
+                        head = op;
+                    }
+                }
+            }
+        );
+        auto is_legal = pos && head;
+        if (!is_legal) {
+            return mlir::failure();
+        }
+
+        auto op = for_op.getOperation();
+        auto pos_clone = rewriter.clone(*pos);
+        auto head_clone = rewriter.clone(*head);
+        auto pos_res = pos_clone->getResult(0);
```

Comment:
This var has one usage, I would just use the expresion in-place. Yes it will mean you will need to split the callsite to multiple lines, but such is life (I think overall it will be a net positive since you will get rid of some local vars, that take up a lot of headspace when reading).